// plumber/lib/plumber.js
require(__dirname + '/../../sinful/sinful.js')
void function(root){

    var plumber, luma, num, pns = [];

    num = require('nmbr')
    luma = require('luma')
    spce = require('spce')

    function riser(to){
        var a = [], i;
        for ( i=0; i<to-1; i++ ){ a[i] = num(0) }
        a[i] = num(1)
        return a
    }

    function m(base, exp){
        var ret;
        if( base.typeLvl < 2 ){
            ret = num(Math.pow(base.value, exp.value))
            return ret
        }else if ( base.typeLvl === 2 ){
            ret = num(num(Math.pow(base.numerator, exp.value)),num(Math.pow(base.denominator, exp.value)))
            return ret
        } else {
            throw new Error('unexpected argument')
        }
    }

    function hashify(input) {
        if ( Array.isArray(input) ) {
            return '['+input.map(function(x){return num(x)}).join(',')+']'
        } else {
            return '['+num(input)+']'
        }
    }

    function plumber(input){
        if (input == null) { throw new Error('missing argument') }

        key = hashify(input)

        if  ( typeof pns[key] === 'undefined' ) {
            pns[key] = PolyInt(input)
        }

        return pns[key]
    }

    var PolyInt = luma.factory({
        init: function(input) {
            var n;
            if(Array.isArray(input)){
                this.ixs = input.map(function(x){return num(x)})
            }else{
                this.ixs = [num(input)]
            }
            this.degree = this.ixs.length - 1
            return this
        }
        , ixs : []
        , typeLvl : 4
        , degree: 0
        , toString: function () { return '['+this.ixs.join(',')+']' }
        , plus : function(input) {
            if (typeof input === 'undefined') throw new Error('argument required')
            if (typeof input.typeLvl === 'undefined') input = num(input)
            if (input.typeLvl < 3 ) input = plumber(input)
            var len = this.ixs.length > input.ixs.length ? this.ixs.length : input.ixs.length
                , i, l, r , result = [];
            for ( i = 0; i < len; i++ ) {
                l = typeof this.ixs[i] !== 'undefined' ? this.ixs[i] : num(0)
                r = typeof input.ixs[i] !== 'undefined' ? input.ixs[i] : num(0)
                result[i] = l.plus(r)
            }
            return plumber(result)
        }
        , minus : function(input) {
            if (typeof input === 'undefined') throw new Error('argument required')
            if (typeof input.typeLvl === 'undefined') input = num(input)
            if (input.typeLvl < 3 ) input = plumber(input)
            var len = this.ixs.length > input.ixs.length ? this.ixs.length : input.ixs.length
                , i, l, r , result = [];
            for ( i = 0; i < len; i++ ) {
                l = typeof this.ixs[i] !== 'undefined' ? this.ixs[i] : num(0)
                r = typeof input.ixs[i] !== 'undefined' ? input.ixs[i] : num(0)
                result[i] = l.minus(r)
            }
            return plumber(result)
        }
        , times : function(input) {
            if (typeof input === 'undefined') throw new Error('argument required')
            if (typeof input.typeLvl === 'undefined') input = num(input)
            if (input.typeLvl < 3 ) input = plumber(input)
            var p = this.ixs, q = input.ixs , result = [] , i, j;
            for ( i=0; i<p.length; i++ ){
                for ( j=0; j<q.length; j++ ){
                    if ( typeof result[i+j] === 'undefined' ) result[i+j] = num(0)
                    result[i+j] = result[i+j].plus(p[i].times(q[j]))
                }
            }
            return plumber(result)
        }
        , pow : function (exp){
            if ( typeof exp === 'undefined' ) throw new Error('argument required')
            if ( typeof exp.typeLvl === 'undefined' ) exp = num(exp)
            if ( exp.typeLvl > 1 ) throw new Error('undefined operation, look for roots elsewhere')
            var result = this, i=0, len=exp.value;
            while ( ++i < exp.value ) {
                result = result.times(this)
            }
            return result
        }
        , val: function(input) {
            var value = num(0), i, len;
            if ( input == null ) throw new Error('missing argument')
            if ( input.typeLvl == null || input.typeLvl < 3 ) input = plumber(input)
            len = this.ixs.length
                debugger
            for ( i=0; i < len; i++ ) {
                value = value.plus(this.ixs[i].times(input.pow(i)))
            }
            return value
        }
    })



    // export
    if ( typeof module != 'undefined' && module.exports )
        module.exports = plumber
    else
        root.factory = plumber

}(this)
